tutorials/plugins/gdnative/gdnative-cpp-scons.rst

```diff
@@ -0,0 +1,872 @@
+        if os.path.isdir(path):
+            add_sources(source_files, path)
+
+This way, we can provide a list of directories to search within for *.cpp files.
+
+Finding Headers
+---------------
+
+Now that we have a means of collecting source files, the last step is to acquire our header files. For this, scons just needs us to supply the directories to search through. It works for both *.h and *.hpp files. These will be "the directories in which we search for all content referred to by the included *.cpp source files".
+
+Right off the bat, we know we'll need to include the ``godot_headers`` directory as well as the generated contents of ``include`` and ``include/core`` from the ``godot-cpp`` directory. ``godot_headers`` is easy to find as it is a submodule within ``godot-cpp``, but the latter we won't necessarily know where to find. As such, we'll need the user to direct us to its location. In addition, we'll give the user the option of specifying an alternate location for the ``godot_headers`` directory as well (helps support legacy versions of ``godot-cpp``).
+
+.. code-block:: python
+    import os
+
+    def add_sources(sources, directory):
+        for file in os.listdir(directory):
+            if file.endswith('.cpp'):
+                sources.append(directory + '/' + file)
+
+    sources = ARGUMENTS.get("sources", "")
+
+    source_dirs = sources.split(",")
+
+    source_files = []
+    for path in source_dirs:
+        if os.path.isdir(path):
+            add_sources(source_files, path)
+    
+    # new content
+    cpp_bindings_path = ARGUMENTS.get("cpp_bindings_path", "godot-cpp/")
+    godot_headers_path = ARGUMENTS.get("headers", cpp_bindings_path + "godot_headers/")
+
+    source_dirs.append(godot_headers_path)
+    source_dirs.append(cpp_bindings_path + "include/")
+    source_dirs.append(cpp_bindings_path + "include/core/")
+
+    # At this point source_dirs has all source and header file directories we reference anywhere.
+    # These directories will be passed to our environment so that scons can locate referenced files.
+
+    env.Append(CPPPATH=source_dirs)
+
+However, there is one piece missing here. Why did we collect the ``source_files`` array, but not ultimately do anything with it? Where does that go?
+
+If you recall, while the environment needs the file locations of the libraries, and the directories of all of our source and header files, the actual source file locations are needed by the dynamic library itself, to know what to include in it. As such, we pass the ``source_files`` array directly to our ``SharedLibrary()`` method.
+
+.. code-block:: python
+
+    library = env.SharedLibrary(target=final_lib_path, source=source_files)
+    Default(library)
+
+Combining Sources, Headers, and Libraries, and Flags
+----------------------------------------------------
+
+Just to review where we are at, let's now pull everything together one more time before moving forward. We're almost at the end!
```

Comment:
We're almost at the end! Here's the complete `SConstruct` file we've created so far: